### PR TITLE
[ROCm][DSv3.2] Eliminate per-decode FillFunctor launches in sparse-MLA hot loop

### PR DESCRIPTION
**Pull request by frida-andersson**
## Summary

On the ROCm sparse-MLA decode path, two `FillFunctor<T>` kernels fire on every indexer-bearing decode step with no observable output — pure launch + bandwidth overhead:

1. `FillFunctor<float>` — `out_logits.fill_(float("-inf"))` in `rocm_fp8_paged_mqa_logits`. Upstream added a persistent `current_workspace_manager()` workspace for `out_logits` (pre-filled once on allocation), but left the per-call `fill_` in place. The downstream `topKPerRowDecode` consumers read only `row[i, :seq_lens[i]]`; the full-row fill was always dead-code.
2. `FillFunctor<int>` — `topk_indices_buffer[:N] = -1` in both sparse-attn-indexer entry points. The decode and prefill kernels fully overwrite active rows and write `-1` to unused topk slots themselves ([sampler.cu:400](https://github.com/vllm-project/vllm/blob/284e6f543/csrc/libtorch_stable/sampler.cu#L400)).

## Changes

`vllm/v1/attention/ops/rocm_aiter_mla_sparse.py`:
- Drop `out_logits.fill_(float("-inf"))` — workspace is pre-filled on allocation by upstream's `current_workspace_manager()`.
- Drop `topk_indices_buffer[:N] = -1` in `rocm_aiter_sparse_attn_indexer`.

`vllm/model_executor/layers/sparse_attn_indexer.py`:
- Drop `topk_indices_buffer[:N] = -1` in `sparse_attn_indexer`.

Pure Python; no HIP/AITER changes.

## Performance — DSv3.2 TP=4 MI355X, 5k ISL / 100 OSL / mc=16

Profiler diff over a 25 s steady-state window:

| Kernel | Before | After | Δ |
|---|---|---|---|
| `FillFunctor<float>` (out_logits fill) | 187,941 calls / 1,153 ms | 0 / 0 ms | **−100%** |
| `FillFunctor<int>` (topk pre-fill) | 193,087 calls / 849 ms | 0 / 0 ms | **−100%** |
| Decode hot-loop wall | 42.27 µs | 23.60 µs | **−44%** |

Aggregate: ~2.0 s of GPU time reclaimed per 25 s window.

Before:
Screenshot 2026-06-04 at 12 25 56
After:
Screenshot 2026-06-04 at 12 26 01

## Correctness — GSM8K 20-shot, `num_fewshot=20`, `num_concurrent=256`

MI355X TP=4, rebased on upstream `main`:

| KV cache | strict-match | flexible-extract |
|---|---|---|
| FP8 | 0.9492 ± 0.006 | 0.9484 ± 0.006 |

Within stderr of published baselines.

## Testing

- [x] GSM8K 20-shot accuracy gate (FP8 KV cache, post-rebase)
- [x] Profiler trace — both `FillFunctor` kernels absent from decode hot loop
- [x] Server boots cleanly from upstream `main` + only these 2 commits
- [x] Zero failed requests across multiple runs

### CODE REVIEW

File: vllm/model_executor/layers/sparse_attn_indexer.py [hunk 1]
```diff
@@ -172,7 +172,6 @@ def sparse_attn_indexer(
             scale_fmt,
         )
 
-    topk_indices_buffer[: hidden_states.shape[0]] = -1
```
Comment (tjtanaa) [inline vllm/model_executor/layers/sparse_attn_indexer.py]: Will have to ensure that this does not affect the behaviour on CUDA as their kernel's behaviour can be different from the kernel used by ROCm.

File: vllm/v1/attention/ops/rocm_aiter_mla_sparse.py [hunk 2]
```diff
@@ -23,6 +23,41 @@
     _ON_GFX950 = False
 
 
+# Persistent workspace for ``rocm_fp8_paged_mqa_logits``'s ``out_logits``.
+# Downstream topk reads ``row[i, :seq_lens[i]]`` only, so per-call
+# ``torch.full(..., -inf)`` was a pure launch + bandwidth tax.
+_PAGED_MQA_LOGITS_BUF: torch.Tensor | None = None
+
+
+def _get_paged_mqa_logits_workspace(
```
Comment (tjtanaa) [inline vllm/v1/attention/ops/rocm_aiter_mla_sparse.py]: @frida-andersson there is a workspace manager that we can use to do this `from vllm.v1.worker.workspace import current_workspace_manager` with `get_simultaneous()`.

File: vllm/model_executor/layers/sparse_attn_indexer.py (hunk 1 above)
Comment (frida-andersson) [reply vllm/model_executor/layers/sparse_attn_indexer.py]: Checked [sampler.cu](https://github.com/vllm-project/vllm/blob/284e6f543/csrc/libtorch_stable/sampler.cu#L400) where topKPerRowDecode/topKPerRowPrefill write -1 directly to unused outIndices slots themselves so the pre-fill should be redundant on both paths

**Comment by frida-andersson** [comment]
Thanks for the review @tjtanaa! The `_PAGED_MQA_LOGITS_BUF` globals are gone since upstream already added `current_workspace_manager()` for `out_logits` independently (exactly what you suggested). After rebasing, commit 1 is now just a 1-line deletion: the `out_logits.fill_(float("-inf"))` call that was still firing every decode step even with the workspace in place

**Comment by dllehr-amd** [comment]
I enabled the ready flag to see if we can see if any fallout from Unit tests can see if the fillfunctor is going to cause a pain somewhere else.

**Comment by frida-andersson** [comment]
Thanks for the ready-flag @dllehr-amd! I updated the branch yesterday and now CI is green

**Comment by frida-andersson** [comment]
Hi @tjtanaa and @dllehr-amd, do you have additional comments on this or can we get it merged as CI is green?

File: vllm/model_executor/layers/sparse_attn_indexer.py (hunk 1 above)
Comment (AndreasKaratzas) [inline vllm/model_executor/layers/sparse_attn_indexer.py]: @zyongye PTAL

**Comment by frida-andersson** [comment]
Rebased onto latest main and re-validated - PTAL @AndreasKaratzas @tjtanaa @dllehr-amd.

Scope shrank after rebase: dropped the sparse_attn_indexer.py hunk (upstream now gates it via skip_topk_buffer_clear), leaving just the two redundant pre-fills in rocm_aiter_mla_sparse.py.

Validated on DSv3.2 TP4, fp8 KV cache, gfx950: GSM8K 20-shot parity (0.9447/0.9431 vs 0.9469/0.9447, within stderr) and the ~440ms of per-layer FillFunctor launches gone from the trace

**Review by tjtanaa — APPROVED** [review]
